modules/mono/editor/GodotTools/GodotTools/Build/DotNetFinder.cs

```diff
@@ -11,8 +11,55 @@ namespace GodotTools.Build
 {
     public static class DotNetFinder
     {
-        public static string? FindDotNetExe()
+        private static string DOTNET_ROOT_ENVIRONMENT_VARIABLE = "DOTNET_ROOT";
+
+        public static string? FindDotNetExe(string? overrideDotnetExecutable)
         {
+            if (string.IsNullOrEmpty(overrideDotnetExecutable) == false)
+            {
+                if (File.Exists(overrideDotnetExecutable))
+                {
+                    return overrideDotnetExecutable;
+                }
+            }
+            // Simplified parallel to modules/mono/editor/hostfxr_resolver.cpp's get_dotnet_root_from_env where
+            // DOTNET_ROOT is checked before path.
+            // https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-environment-variables#net-sdk-and-cli-environment-variables
+            // Further support would be to handle DOTNET_ROOT, DOTNET_ROOT(x86), DOTNET_ROOT_X86, DOTNET_ROOT_X64 for specific architectures.
+            string envVariableToDotnetPath = Environment.GetEnvironmentVariable(DOTNET_ROOT_ENVIRONMENT_VARIABLE);
```

Comment:
I can remove this if you feel strongly about it and in my case, if this goes in, I can still fix my case by using both the DOTNET_ROOT environment variable for the resolver in c++ and the editor variable for the c#.

But there are examples both in the documentation and tools in https://github.com/dotnet/ showing where programs are using DOTNET_ROOT to find the dotnet executable along with the libraries (though most cases are doing it for the libraries).

Some example of tools setting it (with comments about being able to find self):
https://github.com/dotnet/efcore/blob/af2ef93ebbdf0ea74ac988f1f932758ade83d7c8/activate.sh#L44
https://github.com/dotnet/core/blob/b23ca2cbdd8dcbf049b5389e54e194a84e4454c1/release-notes/9.0/install-linux.md?plain=1#L62

Here is an example of using DOTNET_ROOT as the path to the dotnet executable:
https://github.com/dotnet/android/blob/aa668a5c746a7669a145e63efe59d0964093e604/build-tools/scripts/msbuild.mk#L27
https://github.com/dotnet/dotnet-monitor/blob/ee676797d7585abd52d4fed0c9c65646db60d149/startvscode.cmd#L21
https://github.com/dotnet/aspnetcore/blob/7232f994a4f6422d4726a26400c9ac8ed5fa8ca6/src/Hosting/Server.IntegrationTesting/src/Common/DotNetCommands.cs#L74

I can find more examples if that would be useful, but if you feel like you just don't to add this support, I will just strip it in order to get the rest of these changes in.

Reply:
Thanks for the examples, I still think `DOTNET_ROOT` is not the right way to find the location of the dotnet CLI. As I mentioned in the [dev chat](https://chat.godotengine.org/channel/dotnet), `DOTNET_ROOT` is set to the location of the .NET runtime, and not necessarily the location of the dotnet CLI, although they often match.

> https://github.com/dotnet/efcore/blob/af2ef93ebbdf0ea74ac988f1f932758ade83d7c8/activate.sh#L44

This is still setting `PATH`, it's likely also setting `DOTNET_ROOT` so the found runtime matches the dotnet CLI from `PATH`.

> https://github.com/dotnet/core/blob/b23ca2cbdd8dcbf049b5389e54e194a84e4454c1/release-notes/9.0/install-linux.md?plain=1#L62

This says setting `DOTNET_ROOT` is to allow finding the runtime if it's not installed in its regular location, not the dotnet CLI. And then says that the `PATH` environment variable must also be updated to find the dotnet CLI if you don't want to use absolute paths.

> https://github.com/dotnet/android/blob/aa668a5c746a7669a145e63efe59d0964093e604/build-tools/scripts/msbuild.mk#L27

This is setting both the `DOTNET_ROOT` and the path to the dotnet CLI, likely so they match the one built locally in the `/bin` directory of the repo.

> https://github.com/dotnet/dotnet-monitor/blob/ee676797d7585abd52d4fed0c9c65646db60d149/startvscode.cmd#L21

This also sets both `DOTNET_ROOT` and `PATH`.

> https://github.com/dotnet/aspnetcore/blob/7232f994a4f6422d4726a26400c9ac8ed5fa8ca6/src/Hosting/Server.IntegrationTesting/src/Common/DotNetCommands.cs#L74

This tries to find the dotnet CLI in multiple locations, including the local installation directory `.dotnet`, and the location specified by one of these environment variables: `DOTNET_HOME`, `DOTNET_ROOT`, `DOTNET_INSTALL_DIR`, `USERPROFILE`, and `HOME`. But it seems like a rare exception.

Some of these environment variables are OS-specific, and others I couldn't even find documentation for. It's likely they're just trying to be exhaustive to ensure the build scripts work on CI where you may not have a standard installation, although I'd expect this to not be needed if you use the official install script.

---

I'm not convinced by these examples that setting `DOTNET_ROOT` alone should be enough for tools to find the dotnet CLI, since all of them still set the `PATH` environment variable. I'm pretty sure Microsoft's recommendation is to set both `DOTNET_ROOT` and `PATH` when you have .NET installed in a non-standard location. And if you do that, Godot should find the right dotnet CLI since we already look in `PATH`.

A lot of these examples are just customizing these environment variables in order to use the .NET runtime and dotnet CLI from the version locally installed in the `.dotnet` directory. This is a local installation created by the .NET install script in these repos so that they install and use a version of .NET that matches [`global.json`](https://github.com/dotnet/runtime/blob/def360f3275d8c1f26100fdf884fa4d67826e01d/global.json#L3). This is often required in the .NET repos because they use the latest version of .NET including previews, the BCL must be built with the version of .NET that it ships with so they can use the latest features in the codebase.

Also, I think the fact that these scripts exist indicate that you are expected to create your own script to set the `DOTNET_ROOT` and `PATH` environment variables to build with your desired version of .NET. One of the scripts you linked is to launch VSCode after setting these variables, to ensure VSCode is using their preferred .NET installation, similarly you could have a custom script to launch Godot like this:

```bash
DOTNET_ROOT=".dotnet"
PATH=$DOTNET_ROOT:$PATH
godot -e
```

Personally I think Godot using the dotnet CLI to build/publish your project is equivalent to an IDE doing the same, so I feel like rather than looking at custom scripts, we should match what IDEs do. And since one of the scripts you linked is setting these variables before launching VSCode, it sounds to me like VSCode only looks at `PATH` by default.

Reply:
We talked in this thread: https://chat.godotengine.org/channel/dotnet?msg=JPygL2EY3S5GJ8hEJ and I've removed the DOTNET_ROOT support from the c# side and just re-ordered the path searching to not use the hard coded locations until after its checked path and not found it.